File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    let time = 0;
    let bridge = new Array(bridge_length).fill(0);
    let current_weight = 0;
    
    let truck_idx = 0;

    while (truck_idx < truck_weights.length) {
        time++;
        
        let leaving_truck = bridge.shift(); 
        current_weight -= leaving_truck;
        
        if (current_weight + truck_weights[truck_idx] <= weight) {
            let new_truck = truck_weights[truck_idx];
            bridge.push(new_truck);
            current_weight += new_truck;
            truck_idx++;
        } else {
            bridge.push(0);
        }
    }
    
    return time + bridge_length;
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    time = 0
    bridge = deque([0] * bridge_length)
    truck_weights = deque(truck_weights)
    current_weight = 0
    
    while truck_weights:
        time += 1
        leaving_truck = bridge.popleft()
        current_weight -= leaving_truck
        
        if current_weight + truck_weights[0] <= weight:
            new_truck = truck_weights.popleft()
            bridge.append(new_truck)
            current_weight += new_truck
        else:
            bridge.append(0)
            
    return time + bridge_length